In [52]:
import pandas as pd

In [53]:
honeypot = True

pair_wise_out = pd.read_csv("../datasets/Matching Batch 1.csv") 
pair_wise_in = pd.read_csv("../datasets/Matching_HIT.csv")

In [54]:
if honeypot:
    print("Droppin some users")
    pair_wise_out = pair_wise_out[pair_wise_out["WorkerId"] != "A3CDL9X2B4QWOB"]
    pair_wise_out = pair_wise_out[pair_wise_out["WorkerId"] != "A373U4XYXC178J"]    
    pair_wise_out = pair_wise_out[pair_wise_out["WorkerId"] != "A22B5CLRFQ9EUX"]    
    pair_wise_out = pair_wise_out[pair_wise_out["WorkerId"] != "A2236R33CMX70Q"]  
    

Droppin some users


In [55]:
pair_wise_out["Input.C0"]

1      Case Logic 15.6-Inch Laptop and iPad Slim Case...
2      Case Logic 15.6-Inch Laptop and iPad Slim Case...
3      Case Logic 15.6-Inch Laptop and iPad Slim Case...
8      2015 Newest Edition Dell Inspiron 17 5000 Seri...
9      2015 Newest Edition Dell Inspiron 17 5000 Seri...
                             ...                        
129    Lenovo ThinkPad E550 15.6 Inch  i3-4005U 8GB 5...
132    Lenovo ThinkPad E550 15.6 Inch  i3-4005U 8GB 5...
133    Case Logic VLS-114 13-Inch, 14-Inch Laptop Sle...
134    Case Logic VLS-114 13-Inch, 14-Inch Laptop Sle...
137    Case Logic VLS-114 13-Inch, 14-Inch Laptop Sle...
Name: Input.C0, Length: 63, dtype: object

In [56]:
(pair_wise_out["Input.C0"].value_counts())

HP Pavilion 17-g140nr 17.3-Inch Laptop                                                                                                                                             4
Newest Dell Inspiron 14 Inch Laptop with Intel Pentium N3540 Processor up to 2.66 GHz, 4GB DDR3 RAM, 500 GB Hard Drive, No DVD/CD Drive, Windows 8.1 (Certified Refurbished)       4
Case Logic LAPS-114 14-Inch Laptop Sleeve (Black)                                                                                                                                  4
Case Logic VLS-114 13-Inch, 14-Inch Laptop Sleeve (Black)                                                                                                                          3
Kensington 11-Inch Laptop Chromebook Sleeve (K62609WW)                                                                                                                             3
Dell Latitude D820 Laptop Notebook - Core Duo 1.6GHz - 2GB DDR2 - 60GB - DVD+CDRW - Windows 8.1

In [57]:
true_labels_strings=[]
rows=list(pair_wise_out["Input.C0"].value_counts().index)
print(rows)
for i in rows:
    df=pair_wise_in.loc[pair_wise_in["C0"]==i]
#     print(list(df["Labels"].values[0]))
    true_labels_strings.append(df["Labels"].values[0])

['HP Pavilion 17-g140nr 17.3-Inch Laptop', 'Newest Dell Inspiron 14 Inch Laptop with Intel Pentium N3540 Processor up to 2.66 GHz, 4GB DDR3 RAM, 500 GB Hard Drive, No DVD/CD Drive, Windows 8.1 (Certified Refurbished)', 'Case Logic LAPS-114 14-Inch Laptop Sleeve (Black)', 'Case Logic VLS-114 13-Inch, 14-Inch Laptop Sleeve (Black)', 'Kensington 11-Inch Laptop Chromebook Sleeve (K62609WW)', 'Dell Latitude D820 Laptop Notebook - Core Duo 1.6GHz - 2GB DDR2 - 60GB - DVD+CDRW - Windows 8.1 32bit - (Certified Refurbished)', 'OGIO Renegade Laptop Briefcase', 'Case Logic PLS-13 Neoprene 13.3-Inch Neoprene Laptop Sleeve (Black)', 'Case Logic 15.6-Inch Laptop and iPad Slim Case (AUA-316ANTH)', 'Lenovo ThinkPad E550 15.6 Inch  i3-4005U 8GB 500GB SSD Windows 7 Pro Student Notebook Computer', 'Dell Latitude E5500 Laptop Computer - Core 2 Duo 2.0GHz - 2GB DDR2 - 120GB - DVDRW - Windows 7 Pro 64bit - (Certified Refurbished)', 'Case Logic Berkeley Plus BPCA-115 15-Inch Laptop Backpack (Anthracite)', 'Co

In [58]:
true_labels_strings

['[(3, 6)]',
 '[(1, 9), (2, 5)]',
 '[(0, 7), (1, 6), (3, 8)]',
 '[(0, 8), (1, 7), (2, 5)]',
 '[]',
 '[(3, 7), (4, 6)]',
 '[(0, 9), (1, 5), (3, 8)]',
 '[(0, 5), (1, 9), (2, 6), (4, 7)]',
 '[(0, 7), (1, 6), (4, 8)]',
 '[(2, 8)]',
 '[(1, 7), (2, 8), (3, 9)]',
 '[(2, 6)]',
 '[(1, 8), (2, 5), (4, 9)]',
 '[(0, 5), (1, 8), (4, 9)]',
 '[(1, 7), (2, 5)]',
 '[(0, 9), (1, 8), (2, 5)]',
 '[(2, 7), (3, 9), (4, 6)]',
 '[(0, 5), (1, 8), (2, 9)]',
 '[(1, 5), (2, 8), (3, 6)]',
 '[(0, 7), (1, 8), (2, 9), (3, 5)]']

In [59]:
def getLabelsFromString(label_string: str, empty_list: list) -> list:  
    
    if '(' not in label_string:
        return empty_list
    
    first_label = label_string.split('(')[1].split(')')[0]
    rest = label_string.split(')', 1)[1]
    
#     print(rest)
    
    first_element = int(first_label.split(",")[0])
    second_element = int(first_label.split(",")[1])
    element = (first_element, second_element)
    
    empty_list.append(element)
    
    return getLabelsFromString(rest, empty_list)

In [60]:
true_labels = []
for string in true_labels_strings:
    empty = []
    true_labels.append(getLabelsFromString(string, empty))

print(true_labels)

[[(3, 6)], [(1, 9), (2, 5)], [(0, 7), (1, 6), (3, 8)], [(0, 8), (1, 7), (2, 5)], [], [(3, 7), (4, 6)], [(0, 9), (1, 5), (3, 8)], [(0, 5), (1, 9), (2, 6), (4, 7)], [(0, 7), (1, 6), (4, 8)], [(2, 8)], [(1, 7), (2, 8), (3, 9)], [(2, 6)], [(1, 8), (2, 5), (4, 9)], [(0, 5), (1, 8), (4, 9)], [(1, 7), (2, 5)], [(0, 9), (1, 8), (2, 5)], [(2, 7), (3, 9), (4, 6)], [(0, 5), (1, 8), (2, 9)], [(1, 5), (2, 8), (3, 6)], [(0, 7), (1, 8), (2, 9), (3, 5)]]


In [61]:
worker_outputs=[]
for i in rows:
    df=pair_wise_out.loc[pair_wise_out["Input.C0"]==i]
    outputs = [[i,j,m,n,l] for i,j,m,n,l in zip(df["Answer.item0"],df["Answer.item1"],df["Answer.item2"],df["Answer.item3"],df["Answer.item4"])]
    worker_outputs.append(outputs)

In [62]:
len(worker_outputs[3])

3

In [63]:
worker_labels=worker_outputs

In [64]:
import numpy as np
np.array(worker_labels[0])

array([['None', 'None', 'None', 'None', 'None'],
       ['None', 'None', 'None', 'None', 'None'],
       ['5', 'None', 'None', 'None', 'None'],
       ['None', 'None', 'None', 'None', 'None']], dtype='<U4')

In [65]:
pd.Series(np.array(worker_labels[1])[:,0]).value_counts().index[0]

'None'

In [66]:
# Majority voting with 7.
single_label_7=[]
for i in (worker_labels):
    single_label=[]
    mat=np.array(i)
    for k in range(mat.shape[1]):
        f=mat[:,k]
        single_label.append(pd.Series(f).value_counts().index[0])
    single_label_7.append(single_label)
    

In [67]:
(single_label_7)

[['None', 'None', 'None', 'None', 'None'],
 ['None', 'None', '1', 'None', 'None'],
 ['3', 'None', 'None', '4', 'None'],
 ['None', 'None', 'None', 'None', 'None'],
 ['None', 'None', 'None', 'None', 'None'],
 ['None', 'None', 'None', 'None', 'None'],
 ['None', '1', 'None', '4', 'None'],
 ['None', 'None', 'None', 'None', 'None'],
 ['3', '2', '1', 'None', 'None'],
 ['3', 'None', '4', 'None', 'None'],
 ['None', '3', 'None', 'None', 'None'],
 ['None', 'None', 'None', 'None', 'None'],
 ['None', 'None', '1', 'None', 'None'],
 ['1', '4', 'None', 'None', '5'],
 ['None', '3', 'None', 'None', 'None'],
 ['5', 'None', '1', 'None', 'None'],
 ['None', 'None', 'None', 'None', 'None'],
 ['None', 'None', 'None', 'None', 'None'],
 ['None', 'None', 'None', 'None', 'None'],
 ['3', '4', '5', '1', 'None']]

In [68]:
single_label_7[0]

['None', 'None', 'None', 'None', 'None']

In [69]:
converted_labels=[]
for i in single_label_7:
    single_label_c=[]
    for k in enumerate(i):
        if k[1]!="None":
            single_label_c.append((k[0],4+int(k[1])))
    converted_labels.append(single_label_c)

In [70]:
converted_labels

[[],
 [(2, 5)],
 [(0, 7), (3, 8)],
 [],
 [],
 [],
 [(1, 5), (3, 8)],
 [],
 [(0, 7), (1, 6), (2, 5)],
 [(0, 7), (2, 8)],
 [(1, 7)],
 [],
 [(2, 5)],
 [(0, 5), (1, 8), (4, 9)],
 [(1, 7)],
 [(0, 9), (2, 5)],
 [],
 [],
 [],
 [(0, 7), (1, 8), (2, 9), (3, 5)]]

In [71]:
allposs=[]
for i in range(5):
    for j in range(5,10):
        allposs.append((i,j))
cms=[]
precs=[]
recs=[]
for i,j in zip(true_labels,converted_labels):
    TP= len(i)-len(set(i)-set(j))
    FP= len(set(j)-set(i))
    TN= -1*((len(set(allposs)))-(len(set(allposs)-set(i)))-len(set(allposs)-set(j)))
    FN= len(allposs)-TP-FP-TN
    print("TP,FP,TN,FN are as follows for this question.")
    print(i)
    print(j)
    cm=[TP,FP,TN,FN]
    cms.append(cm)
    print(cm)
    

TP,FP,TN,FN are as follows for this question.
[(3, 6)]
[]
[0, 0, 24, 1]
TP,FP,TN,FN are as follows for this question.
[(1, 9), (2, 5)]
[(2, 5)]
[1, 0, 22, 2]
TP,FP,TN,FN are as follows for this question.
[(0, 7), (1, 6), (3, 8)]
[(0, 7), (3, 8)]
[2, 0, 20, 3]
TP,FP,TN,FN are as follows for this question.
[(0, 8), (1, 7), (2, 5)]
[]
[0, 0, 22, 3]
TP,FP,TN,FN are as follows for this question.
[]
[]
[0, 0, 25, 0]
TP,FP,TN,FN are as follows for this question.
[(3, 7), (4, 6)]
[]
[0, 0, 23, 2]
TP,FP,TN,FN are as follows for this question.
[(0, 9), (1, 5), (3, 8)]
[(1, 5), (3, 8)]
[2, 0, 20, 3]
TP,FP,TN,FN are as follows for this question.
[(0, 5), (1, 9), (2, 6), (4, 7)]
[]
[0, 0, 21, 4]
TP,FP,TN,FN are as follows for this question.
[(0, 7), (1, 6), (4, 8)]
[(0, 7), (1, 6), (2, 5)]
[2, 1, 19, 3]
TP,FP,TN,FN are as follows for this question.
[(2, 8)]
[(0, 7), (2, 8)]
[1, 1, 22, 1]
TP,FP,TN,FN are as follows for this question.
[(1, 7), (2, 8), (3, 9)]
[(1, 7)]
[1, 0, 21, 3]
TP,FP,TN,FN are as

In [72]:
FinalCM=np.array(cms).mean(axis=0)
print(FinalCM)

[ 1.   0.1 21.4  2.5]


In [73]:
Precision = FinalCM[0]/(FinalCM[0]+FinalCM[1])

In [74]:
Recall = FinalCM[0]/(FinalCM[0]+FinalCM[3])

In [75]:
Precision

0.9090909090909091

In [76]:
Recall

0.2857142857142857

In [77]:
F1_score = (2 * Precision * Recall) / (Precision + Recall)

In [78]:
F1_score

0.43478260869565216

## WITH NO HONEYPOT
#### Precision: 0.929
#### Recall: 0.206
#### F1 score: 0.338

## WITH HONEYPOT
#### Precision: 0.9090
#### Recall: 0.286
#### F1 score: 0.435